processing/src/main/java/io/druid/segment/StringDimensionHandler.java

```diff
@@ -0,0 +1,1248 @@
+  {
+    private final List<String> sortedVals;
+    private final int[] idToIndex;
+    private final int[] indexToId;
+
+    public SortedDimensionDictionary(List<String> idToValue, int length)
+    {
+      Map<String, Integer> sortedMap = Maps.newTreeMap();
+      for (int id = 0; id < length; id++) {
+        sortedMap.put(idToValue.get(id), id);
+      }
+      this.sortedVals = Lists.newArrayList(sortedMap.keySet());
+      this.idToIndex = new int[length];
+      this.indexToId = new int[length];
+      int index = 0;
+      for (Integer id : sortedMap.values()) {
+        idToIndex[id] = index;
+        indexToId[index] = id;
+        index++;
+      }
+    }
+
+    public int size()
+    {
+      return sortedVals.size();
+    }
+
+    public int getUnsortedIdFromSortedId(int index)
+    {
+      return indexToId[index];
+    }
+
+    public int getSortedIdFromUnsortedId(int id)
+    {
+      return idToIndex[id];
+    }
+
+    public String getValueFromSortedId(int index)
+    {
+      return Strings.emptyToNull(sortedVals.get(index));
+    }
+  }
+
+  public class StringDimensionIndexer implements DimensionIndexer<Integer, String>
+  {
+    private DimensionDictionary dimLookup;
+    private SortedDimensionDictionary sortedLookup;
+    private Object lock;
+
+    public StringDimensionIndexer(Object lock)
+    {
+      this.lock = lock;
+      this.dimLookup = new DimensionDictionary(this.lock);
+    }
+
+    @Override
+    public void addNullLookup()
+    {
+      int oldDictSize = dimLookup.size();
+      dimLookup.add(null);
+      if (oldDictSize != dimLookup.size()) {
+        sortedLookup = dimLookup.sort();
+      }
+    }
+
+    @Override
+    public Comparable[] processRowValsToIndexKey(Object dimValues)
+    {
+      final Integer[] encodedDimensionValues;
+      int oldDictSize = dimLookup.size();
+      try {
+        if (dimValues == null) {
+          dimLookup.add(null);
+          encodedDimensionValues = null;
+        } else if (dimValues instanceof List) {
+          List<Object> dimValuesList = (List) dimValues;
+
+          // Sort multival row by their unencoded values first.
+          final String[] dimensionValues = new String[dimValuesList.size()];
+          for (int i = 0; i < dimValuesList.size(); i++) {
+            dimensionValues[i] = STRING_TRANSFORMER.apply(dimValuesList.get(i));
+          }
+          Arrays.sort(dimensionValues, UNENCODED_COMPARATOR);
+
+          encodedDimensionValues = new Integer[dimensionValues.length];
+          for (int i = 0; i < dimensionValues.length; i++) {
+            encodedDimensionValues[i] = dimLookup.add(dimensionValues[i]);
+          }
+        } else {
+          String transformedVal = STRING_TRANSFORMER.apply(dimValues);
+          encodedDimensionValues = new Integer[]{dimLookup.add(transformedVal)};
+        }
+      }
+      catch (ParseException pe) {
+        throw new ParseException(pe.getMessage() + dimensionName);
```

Comment:
please add original exception to the new constructor, so that original stacktrace is preserved.